src/module/functions/mod.rs

```diff
@@ -206,6 +210,17 @@ impl ModuleFunctions {
         self.arena.delete(id);
     }
 
+    /// Remove a function from this module by name that corresponds to it's import or export.
+    ///
+    /// Unlike [`by_name()`](ModuleFunctions::by_name), "name" corresponds to either a functions import name or it's export name.
+    pub fn delete_by_name(&mut self, name: impl AsRef<str>) {
```

Comment:
Again, perhaps this should be by Id.

Reply:
Right then perhaps delete is superfluous then. Having them on `ModuleFunctions` seems preferable to me. And that also allows them to have shorter names.

Reply:
Thanks for the feedback, will get to integrating this. I filled out more of the other implementations and I need to do some looking and thinking, should have something ready for review tomorrow!

- [ ] Move all func-y modules to `ModuleFunctions`

Reply:
Hey @guybedford what did you think about the `replace_fn_by_id` stuff here? That's the biggest hairiest fn, and it just seems to be a little bit messier than the others.  

If you've got any ideas for refactoring, I'd love to hear them -- it's a little bit messy as-is, I'm not a huge fan of just *trying* to get the function from the imports, as opposed to having two instances of `replace_func_by_id` (one for `ModuleImports` another for `ModuleExports`). 

The unified implementation just ends up being basically two big `if`s that do just-slightly-different things for imports/exports because of how the types/destructuring works out.

`module.exports.replace_func_by_id(...)` and `module.imports.replace_func_by_id(...)` doesn't seem so bad versusa `module.replace_func_by_id(...)` which is implemented as it is now.

Reply:
Yeah I appreciate it's got two completely separate branches here. What happens if the function is not an import or an export, does it work correctly then as well?

Reply:
At present, it throws an error like this:

```rust
bail!("function with ID [{fid:?}] is neither an import or an export");
```

Is that right? I was assuming that we want to fail if a function is neither an import nor export.

Assuming it's "right" It's not terrible, but it's also not *great* either, I think.

The big thing that worried me which was missing was the "upsert" semantic of [the similar downstream function in WASI-virt](https://github.com/bytecodealliance/WASI-Virt/blob/main/src/walrus_ops.rs#L130). Since the original functions were by name, it had the ability to *insert* the function if it was missing.

This means I couldn't perfectly replace that stubbing functionality (but of course, someone could write quite succinct code that checked for the existing function and replaced as necessary, something like:

```rust
if let Some(fid) = module.exports.get_func_by_name(export_name) {
    replace_func_by_id(...);
}  else {
    module.exports.add(export_name, ...) // function builder stuff has to get duplicated here, unfortunately
}
```

So to summarize, it *feels* like:
- the functions should be manipulated by name (so we can preserve that upsert functionality)
- the replacement function should probably live on the individual imports/exports objects

Reply:
Yeah I think this function as framed as a kind of replacer by id, only makes sense if it also works for arbitrary locals? But then perhaps I've framed this wrong and focusing on the separate import and export cases is better?

It would be fine to just wrap a get_or_create_export functionality around it if we could figure out a way to make it work properly though I think for the WASI-Virt case.

I'm open to suggestions though.

Reply:
> It would be fine to just wrap a get_or_create_export functionality around it if we could figure out a way to make it work properly though I think for the WASI-Virt case.

I do like adding a `get_or_create_export` so WASI-Virt can be cleaner... What about a mash of the solutions?

`ModuleExports::replace_func_by_id` (or `replace_func()` ?
`Module::get_or_create_export`

`ModuleImports::replace_func_by_id`
`Module::get_or_create_import`

There's a bit of duplication, but I think it's definitely least-surprise.

Reply:
oh also, I haven't forgotten about tests! Will add after we've figured out a shape that feels reasonable

Reply:
Okay, I'm open to that if you would prefer. I think for this kind of a `replace_func_by_id` to work it would absolutely need to handle functions that aren't exported or imported.

Reply:
Yeah it's not ideal but the reason I'm thinking `replace_func_by_id` (if it worked for local functions) would be troublesome is that if you tried to use it from `get_or_create_import` it *could* be misused and replace a local function by accident.

I'll think about it a bit more -- please let me know if you have any ideas on how to write it in a way that could work for either... 

I mean it's overkill but we could have those function *AND* a `Module::replace_func_by_id` which uses the other two (and handles local functions). Still seems possible to misuse, but we can have the `module.imports.replace_func...` and `module.replace_func...` which works for everything.

Reply:
Yeah, perhaps let's just focus on the separate import and export cases for now.